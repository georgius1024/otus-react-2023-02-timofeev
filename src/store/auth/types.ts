import type { User } from "@/types";

export interface AuthState {
  user: User | null;
}
