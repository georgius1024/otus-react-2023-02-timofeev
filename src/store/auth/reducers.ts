import type { AuthState } from "@/store/auth/types";

export default {
  logout: (state: AuthState) => {
    state.user = undefined;
  },
  store: (state: AuthState) => {
    localStorage["store/auth"] = JSON.stringify(state);
  },
  restore: (state: AuthState) => {
    if ("store/auth" in localStorage) {
      try {
        state = JSON.parse(localStorage["store/auth"]);
        state.busy = undefined
        state.error = undefined
      } catch (error) {
        console.error(error);
      }
    }
  },
};
