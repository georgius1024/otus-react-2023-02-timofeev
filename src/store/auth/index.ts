import { createSlice } from "@reduxjs/toolkit";
import type { AuthState } from "@/store/auth/types";
import reducers from "@/store/auth/reducers";

const initialState: AuthState = {
  user: null,
};

export const authSlice = createSlice({
  name: "store/auth",
  initialState,
  reducers
});

export const { login, logout, store, restore } = authSlice.actions;

export default authSlice.reducer;
