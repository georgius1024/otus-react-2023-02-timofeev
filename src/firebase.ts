import { initializeApp } from "firebase/app";
import { getAuth } from "firebase/auth";
import { getFirestore } from "firebase/firestore";

const firebaseConfig = {
  apiKey: "AIzaSyA71hM5n6OrH1uTWT_TWBJiugO0wibmY9o",
  authDomain: "react-firebase-project-3e66e.firebaseapp.com",
  projectId: "react-firebase-project-3e66e",
  storageBucket: "react-firebase-project-3e66e.appspot.com",
  messagingSenderId: "209900050532",
  appId: "1:209900050532:web:33a3092e817ca26878a771",
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);

export const auth = getAuth(app);
export const db = getFirestore(app);

export default app;
