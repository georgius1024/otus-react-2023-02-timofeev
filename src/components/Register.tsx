import { PureComponent } from "react";
import { createUserWithEmailAndPassword } from "firebase/auth";
import { auth } from "@/firebase";

import type { Navigate, Login } from "@/types";
import { Page } from "@/types";

type RegisterProps = {
  navigate: Navigate;
  login: Login;
};
type RegisterState = {
  email: string;
  password: string;
}

class RegisterPage extends PureComponent<RegisterProps, RegisterState> {
  state: RegisterState = {
    email: "", password: ""
  }
  constructor(props: RegisterProps) {
    super(props);
  }
  setEmail = (event: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({ email: event.target.value });
  }
  setPassword = (event: React.ChangeEvent<HTMLInputElement>): void  => {
    this.setState({ password: event.target.value });
  }
  submit = (event: React.MouseEvent<HTMLButtonElement>): void => {
    event.preventDefault();
    createUserWithEmailAndPassword(auth, this.state.email, this.state.password)
      .then((userCredential) => {
        this.props.login(userCredential.user)
        this.props.navigate(Page.home)
        alert("Thank you for registration");
      })
      .catch((error) => {
        const errorCode = error.code;
        const errorMessage = error.message;
        console.log(errorCode, errorMessage);
        alert("Thank you for registration");
      });
  }
  render() {
    return (
      <form name="register">
        <h1>Register</h1>
        <input
          value={this.state.email}
          type="email"
          name="email"
          onInput={this.setEmail}
        />
        <label>Password</label>
        <input
          value={this.state.password}
          type="password"
          onInput={this.setPassword}
        />
        <a href="javascript:void(0)" onClick={() => this.props.navigate(Page.forgot)}>Forgot password?</a>
        <button type="button" onClick={this.submit}>
          Register
        </button>
      </form>
    );
  }
}

export default RegisterPage;
