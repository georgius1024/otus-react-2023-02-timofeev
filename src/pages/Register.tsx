import { PureComponent } from "react";
import { createUserWithEmailAndPassword } from "firebase/auth";
import { auth } from "@/firebase";
import { Link } from "react-router-dom";
import type { Login } from "@/types";

type RegisterProps = {
  register: Login;
};

type RegisterState = {
  email: string;
  password: string;
}

class RegisterPage extends PureComponent<RegisterProps, RegisterState> {
  state: RegisterState = {
    email: "", password: ""
  }
  constructor(props: RegisterProps) {
    super(props);
  }
  setEmail = (event: React.ChangeEvent<HTMLInputElement>): void => {
    this.setState({ email: event.target.value });
  }
  setPassword = (event: React.ChangeEvent<HTMLInputElement>): void  => {
    this.setState({ password: event.target.value });
  }
  submit = (event: React.MouseEvent<HTMLButtonElement>): void => {
    event.preventDefault();
    createUserWithEmailAndPassword(auth, this.state.email, this.state.password)
      .then((userCredential) => {
        this.props.register(userCredential.user)
        alert("Thank you for registration");
      })
      .catch((error) => {
        const errorCode = error.code;
        const errorMessage = error.message;
        console.log(errorCode, errorMessage);
        alert("Thank you for registration");
      });
  }
  render() {
    return (
      <form name="register">
        <h1>Register</h1>
        <input
          value={this.state.email}
          type="email"
          name="email"
          onInput={this.setEmail}
        />
        <label>Password</label>
        <input
          value={this.state.password}
          type="password"
          onInput={this.setPassword}
        />
        <Link to="/forgot">Forgot password?</Link>
        <button type="button" onClick={this.submit}>
          Register
        </button>
      </form>
    );
  }
}

export default RegisterPage;
