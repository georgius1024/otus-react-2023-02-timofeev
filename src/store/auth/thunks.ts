import {
  createAsyncThunk,
  ActionReducerMapBuilder,
  PayloadAction,
} from "@reduxjs/toolkit";

import {
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  sendPasswordResetEmail,
} from "firebase/auth";
import { auth } from "@/firebase";

import type { AuthState, AuthPayload } from "@/store/auth/types";
import type { User } from "@/types";

export const login = createAsyncThunk<User, AuthPayload, { rejectValue: string }>(
  "auth/login",
  async (payload: AuthPayload) => {
    return signInWithEmailAndPassword(
      auth,
      payload.email,
      payload.password
    ).then((userCredentials) => {
      const { uid, email, providerData } = userCredentials.user;
      return { uid, email, providerData } as User;
    });
  }
);

export const register = createAsyncThunk(
  "auth/register",
  async (payload: AuthPayload) => {
    return createUserWithEmailAndPassword(
      auth,
      payload.email,
      payload.password
    ).then((userCredentials) => {
      const { uid, email, providerData } = userCredentials.user;
      return { uid, email, providerData };
    });
  }
);

export const forgot = createAsyncThunk("auth/forgot", async (email: string) => {
  return sendPasswordResetEmail(auth, email);
});

export default function (builder: ActionReducerMapBuilder<AuthState>) {
  builder
    .addCase(login.pending, (state: AuthState) => {
      state.busy = true;
      state.user = undefined;
      state.error = undefined;
    })
    .addCase(login.fulfilled, (state: AuthState, action) => {
      state.busy = false;
      state.user = action.payload as unknown as User;
    })
    .addCase(login.rejected, (state: AuthState, action) => {
      state.busy = false;
      state.error = action.error.message;
    })
    .addCase(register.pending, (state: AuthState) => {
      state.busy = true;
      state.user = undefined;
      state.error = undefined;
    })
    .addCase(register.fulfilled, (state: AuthState, action) => {
      state.busy = false;
      state.user = action.payload as unknown as User;
    })
    .addCase(register.rejected, (state: AuthState, action) => {
      state.busy = false;
      state.error = action.error.message;
    })
    .addCase(forgot.pending, (state: AuthState) => {
      state.busy = true;
      state.error = undefined;
    })
    .addCase(forgot.fulfilled, (state: AuthState) => {
      state.busy = false;
    })
    .addCase(forgot.rejected, (state: AuthState, action) => {
      state.busy = false;
      state.error = action.error.message;
    });
}
