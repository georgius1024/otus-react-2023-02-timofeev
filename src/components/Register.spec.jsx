import { render, screen, fireEvent } from "@testing-library/react";
import { vi } from "vitest";
import flushPromises from 'flush-promises'

vi.mock("firebase/auth", () => ({
  getAuth: vi.fn().mockResolvedValue(true),
  createUserWithEmailAndPassword: vi.fn().mockResolvedValue(true)
}))

import * as firebaseAuth from "firebase/auth"
import Register from "./Register.tsx";

describe("Register", () => {
  it("renders without errors", () => {
    expect(render(<Register navigate={vi.fn()} login={vi.fn()} />)).toMatchSnapshot();
  });

  it("button click triggers request and navigation", async () => {
    const navigate = vi.fn()
    const login = vi.fn()
    render(<Register navigate={navigate} login={login} />);
    const button = screen.getByRole('button', {
      name: /Register/i
    })
    fireEvent.click(button);
    await flushPromises()
    expect(firebaseAuth.createUserWithEmailAndPassword).toHaveBeenCalled();
    expect(navigate).toHaveBeenCalled();
    expect(login).toHaveBeenCalled();
  });
});
