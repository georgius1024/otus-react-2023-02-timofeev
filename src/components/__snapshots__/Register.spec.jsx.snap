// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Register > renders without errors 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        name="register"
      >
        <h1>
          Register
        </h1>
        <input
          name="email"
          type="email"
          value=""
        />
        <label>
          Password
        </label>
        <input
          type="password"
          value=""
        />
        <a
          href="javascript:void(0)"
        >
          Forgot password?
        </a>
        <button
          type="button"
        >
          Register
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      name="register"
    >
      <h1>
        Register
      </h1>
      <input
        name="email"
        type="email"
        value=""
      />
      <label>
        Password
      </label>
      <input
        type="password"
        value=""
      />
      <a
        href="javascript:void(0)"
      >
        Forgot password?
      </a>
      <button
        type="button"
      >
        Register
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
